### PR TITLE
Pause when in background

### LINKED ISSUE #4481: [Mini Feature Request] Pause when background or when minimized
i know there is a pause feature from emulation tab, but it come in handy  when its pause automatically to pause in certain situation such as another app popup or minimized citra (inspired by mesen emulator)
asd

### PR DESCRIPTION
**Pull request by vitor-k**
Addresses and should close #4481. 

Creates the option for Citra to pause when it's in the background or minimized.

Pause when on background

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/4930)

### CODE REVIEW

File: src/citra_qt/main.cpp [hunk 1]
```diff
@@ -499,6 +499,21 @@ void GMainWindow::RestoreUIState() {
     statusBar()->setVisible(ui.action_Show_Status_Bar->isChecked());
 }
 
+void GMainWindow::OnLoseFocus(Qt::ApplicationState state) {
```
Comment (wwylele) [inline src/citra_qt/main.cpp]: Better rename this to `OnFocusStateChanged` is it also handles gaining focus.

File: src/citra_qt/main.cpp [hunk 2]
```diff
@@ -2013,6 +2028,10 @@ int main(int argc, char* argv[]) {
     Core::System::GetInstance().RegisterSoftwareKeyboard(std::make_shared<QtKeyboard>(main_window));
 
     main_window.show();
+
+    QObject::connect(&app, &QGuiApplication::applicationStateChanged, &main_window,
```
Comment (zhaowenlan1779) [inline src/citra_qt/main.cpp]: The QObject:: seems unnecessary.

File: src/citra_qt/configuration/configure_general.ui [hunk 3]
```diff
@@ -29,6 +29,13 @@
           </property>
          </widget>
         </item>
+       <item>
+         <widget class="QCheckBox" name="toggle_background_pause">
+          <property name="text">
+           <string>Pause emulation when on background</string>
```
Comment (lioncash) [inline src/citra_qt/configuration/configure_general.ui]:
```suggestion
           <string>Pause emulation when in background</string>
```

is the general phrasing for this

File: src/citra_qt/main.cpp [hunk 4]
```diff
@@ -499,6 +499,21 @@ void GMainWindow::RestoreUIState() {
     statusBar()->setVisible(ui.action_Show_Status_Bar->isChecked());
 }
 
+void GMainWindow::OnAppFocusStateChanged(Qt::ApplicationState state) {
+    if (UISettings::values.pause_when_in_background) {
```
Comment (lioncash) [inline src/citra_qt/main.cpp]: Inverting this into a guard clause like:

```suggestion
    if (!UISettings::values.pause_when_in_background) {
        return;
    }
```

will allow you to unindent code one level

**Comment by pbarilla** [comment]
Tried it on my local, works great! It doesn't pause when Mission Control is enabled on macOS, but that makes sense, since it hasn't actually lost focus at that point.

File: src/citra_qt/configuration/configure_general.ui [hunk 5]
```diff
@@ -29,6 +29,13 @@
           </property>
          </widget>
         </item>
+       <item>
```
Comment (zhaowenlan1779) [inline src/citra_qt/configuration/configure_general.ui]:
```suggestion
        <item>
```
To align properly.

Comment (vitor-k) [reply src/citra_qt/configuration/configure_general.ui]: I'll wait a bit, if somebody reports a problem with canary I'll include this in the next commit.
If there are no problems I'll try to push this change without a new commit (and try not to cause a canary rebuild)

**Pull request merged**